File: cups.pas
program CupsProg;
{$mode objfpc}
{$m+}

Uses math, sysutils, contnrs;

type
  CupsPtr = ^Cups;
  StatePtr = ^CupsState;

  Cups = Class
    private
      x, y: Integer;
    public
      constructor new(_x, _y: Integer);
      destructor done();
      function valid(): Boolean;
      function getX(): Integer;
      function getY(): Integer;
      function toString(): String;
      function fillX(): Cups;
      function fillY(): Cups;
      function emptyX(): Cups;
      function emptyY(): Cups;
      function pourXY(): Cups;
      function pourYX(): Cups;
  end;

  CupsState = Class
    private
      steps: Integer;
      parent: CupsPtr;
    public
      constructor new(s: Integer; p: CupsPtr);
      destructor done();
      function getSteps(): Integer;
      function getParent(): CupsPtr;
  end;
var
  pushedNewState: boolean;
  currentCups, nextCups: CupsPtr;
  currentState, nextState: StatePtr;
  hash: TFPHashList;
  queue: TQueue;
  outputStack: TStack;
const
  VOLUME_X = 4;
  VOLUME_Y = 9;
  TARGET_VOLUME = 2;

constructor Cups.new(_x, _y: Integer);
begin
  x := _x;
  y := _y;
end;

destructor Cups.done();
begin
end;

function Cups.valid(): Boolean;
begin
  valid := ((x = TARGET_VOLUME) or (y = TARGET_VOLUME));
end;

function Cups.getX(): Integer;
begin
  getX := x;
end;

function Cups.getY(): Integer;
begin
  getY := y;
end;

function Cups.toString(): String;
begin
  toString := '(' + IntToStr(x) + ', ' + IntToStr(y) + ')';
end;

function Cups.fillX(): Cups;
begin
  fillX := Cups.new(VOLUME_X, y);
end;

function Cups.fillY(): Cups;
begin
  fillY := Cups.new(x, VOLUME_Y);
end;

function Cups.emptyX(): Cups;
begin
  emptyX := Cups.new(0, y);
end;

function Cups.emptyY(): Cups;
begin
  emptyY := Cups.new(x, 0);
end;

function Cups.pourXY(): Cups;
var
  newX, newY: Integer;
begin
  newX := min(VOLUME_X, x + y);
  newY := y - (newX - x);
  pourXY := Cups.new(newX, newY);
end;

function Cups.pourYX(): Cups;
var
  newX, newY: Integer;
begin
  newY := min(VOLUME_Y, x + y);
  newX := x - (newY - y);
  pourYX := Cups.new(newX, newY);
end;

constructor CupsState.new(s: Integer; p: CupsPtr);
begin
  steps := s;
  parent := p;
end;

destructor CupsState.done();
begin
end;

function CupsState.getSteps(): Integer;
begin
  getSteps := steps;
end;

function CupsState.getParent(): CupsPtr;
begin
  getParent := parent;
end;

begin
  new(currentCups);
  new(currentState);
  currentCups^ := Cups.new(0, 0);
  currentState^ := CupsState.new(0, nil);
  hash := TFPHashList.create();
  queue := TQueue.create();

  queue.push(currentCups);
  hash.add(currentCups^.toString(), currentState);

  while queue.count > 0 do
  begin
    currentCups := CupsPtr(queue.pop);
    currentState := StatePtr(hash.find(currentCups^.toString));
    new(nextState);
    nextState^ := CupsState.new(currentState^.getSteps() + 1, currentCups);
    pushedNewState := false;

    if currentCups^.valid then
    begin
      outputStack := TStack.create();
      writeln(currentState^.getSteps());

      while currentCups <> nil do
      begin
        outputStack.push(currentCups);

        currentState := StatePtr(hash.find(currentCups^.toString));
        currentCups := currentState^.getParent();
      end;

      while outputStack.count > 0 do
      begin
        currentCups := CupsPtr(outputStack.pop());
        writeln(currentCups^.toString());
      end;
      break
    end;

    new(nextCups);
    nextCups^ := currentCups^.fillX();
    if hash.find(nextCups^.toString) = nil then
    begin
      queue.push(nextCups);
      hash.add(nextCups^.toString, nextState);
      pushedNewState := true;
    end
    else
    begin
      nextCups^.done();
      dispose(nextCups);
    end;

    new(nextCups);
    nextCups^ := currentCups^.fillY();
    if hash.find(nextCups^.toString) = nil then
    begin
      queue.push(nextCups);
      hash.add(nextCups^.toString, nextState);
      pushedNewState := true;
    end
    else
    begin
      nextCups^.done();
      dispose(nextCups);
    end;

    new(nextCups);
    nextCups^ := currentCups^.emptyX();
    if hash.find(nextCups^.toString) = nil then
    begin
      queue.push(nextCups);
      hash.add(nextCups^.toString, nextState);
      pushedNewState := true;
    end
    else
    begin
      nextCups^.done();
      dispose(nextCups);
    end;

    new(nextCups);
    nextCups^ := currentCups^.emptyY();
    if hash.find(nextCups^.toString) = nil then
    begin
      queue.push(nextCups);
      hash.add(nextCups^.toString, nextState);
      pushedNewState := true;
    end
    else
    begin
      nextCups^.done();
      dispose(nextCups);
    end;

    new(nextCups);
    nextCups^ := currentCups^.pourXY();
    if hash.find(nextCups^.toString) = nil then
    begin
      queue.push(nextCups);
      hash.add(nextCups^.toString, nextState);
      pushedNewState := true;
    end
    else
    begin
      nextCups^.done();
      dispose(nextCups);
    end;

    new(nextCups);
    nextCups^ := currentCups^.pourYX();
    if hash.find(nextCups^.toString) = nil then
    begin
      queue.push(nextCups);
      hash.add(nextCups^.toString, nextState);
      pushedNewState := true;
    end
    else
    begin
      nextCups^.done();
      dispose(nextCups);
    end;

    if not pushedNewState then
    begin
      nextState^.done();
      dispose(nextState);
    end;
  end;
end.

File: cups.go
package main

import "fmt"
import "strconv"
import "container/list"

const (
    VOLUME_X int = 4
    VOLUME_Y int = 9
    TARGET_VOLUME int = 2
)

type Cups struct {
    x int
    y int
}

type CupsState struct {
    steps int
    parent *Cups
}

func (c *Cups) toString() string {
    return "(" + strconv.Itoa(c.x) + ", " + strconv.Itoa(c.y) + ")"
}

func (c *Cups) valid() bool {
    return c.x == TARGET_VOLUME || c.y == TARGET_VOLUME
}

func (c *Cups) fillX() *Cups {
    return &Cups{x: VOLUME_X, y: c.y}
}

func (c *Cups) fillY() *Cups {
    return &Cups{x: c.x, y: VOLUME_Y}
}

func (c *Cups) emptyX() *Cups {
    return &Cups{x: 0, y: c.y}
}

func (c *Cups) emptyY() *Cups {
    return &Cups{x: c.x, y: 0}
}

func (c *Cups) pourXY() *Cups {
    newY := int(min(VOLUME_Y, c.x + c.y))
    newX := c.x - (newY - c.y)
    return &Cups{x: newX, y: newY}
}

func (c *Cups) pourYX() *Cups {
    newX := int(min(VOLUME_X, c.x + c.y))
    newY := c.y - (newX - c.x)
    return &Cups{x: newX, y: newY}
}

func (c *Cups) mutate() [6]*Cups {
    return [6]*Cups{c.fillX(), c.fillY(), c.emptyX(), c.emptyY(), c.pourXY(), c.pourYX()}
}

func min(a, b int) int {
    if a < b {
        return a
    } else {
        return b
    }
}

func main() {
    var currentCups *Cups
    var nextCups [6]*Cups
    var currentState, nextState CupsState
    queue := list.New()
    hash := make(map[Cups]CupsState)

    currentCups = new(Cups)
    queue.PushBack(*currentCups)
    hash[*currentCups] = CupsState{steps: 0, parent: nil}

    for queue.Len() > 0 {
        e := queue.Front()
        queue.Remove(e)
        currentCups = new(Cups)
        *currentCups = e.Value.(Cups)
        currentState = hash[*currentCups]
        nextState = CupsState{steps: currentState.steps + 1, parent: currentCups}

        if currentCups.valid() {
            fmt.Println(currentState.steps)
            outputStack := list.New()
            for currentCups != nil {
                outputStack.PushBack(currentCups.toString())
                currentCups = hash[*currentCups].parent
            }
            for outputStack.Len() > 0 {
                e := outputStack.Back()
                outputStack.Remove(e)
                fmt.Println(e.Value)
            }
            break
        }

        nextCups = currentCups.mutate()
        for _, cups := range nextCups {
            _, exists := hash[*cups]
            if !exists {
                queue.PushBack(*cups)
                hash[*cups] = nextState
            }
        }
    }
}
